batcher.go

```diff
@@ -317,44 +371,48 @@ func (r *Batcher) Start() (err error) {
 	// define the func for flushing a batch
 	var lastFlushWithRecords time.Time
 	flush := func(watcher IWatcher, batch []IOperation) {
-		if len(batch) > 0 {
-			if r.emitBatch {
-				r.emit(BatchEvent, len(batch), "", batch)
+		if len(batch) == 0 {
+			return
+		}
+		if r.emitBatch {
+			r.emit(BatchEvent, len(batch), "", batch)
+		}
+		lastFlushWithRecords = time.Now()
+		go func() {
+
+			// increment an attempt
+			for _, op := range batch {
+				op.MakeAttempt()
 			}
-			lastFlushWithRecords = time.Now()
+
+			// process the batch
+			waitForDone := make(chan struct{})
 			go func() {
+				defer close(waitForDone)
+				watcher.ProcessBatch(batch)
+			}()
 
-				// increment an attempt
-				for _, op := range batch {
-					op.MakeAttempt()
-				}
+			// the batch is "done" when the ProcessBatch func() finishes or the maxOperationTime is exceeded
+			maxOperationTime := r.maxOperationTime
+			if watcher.MaxOperationTime() > 0 {
+				maxOperationTime = watcher.MaxOperationTime()
+			}
+			select {
+			case <-waitForDone:
+			case <-time.After(maxOperationTime):
+			}
```

Comment:
I'm wondering if this would be more clear if structured differently...

if watcher.max... > 0 {
    select {
        case <-waitForDone:
        case <-time.After(watcher.max...):
    }
} else {
    <-waitForDone
}

Reply:
The behavior is actually different than you are showing. The behavior is waitForDone or wait for the maxOperationTime. But maxOperationTime can either come from Batcher (which always has one because there is a default) or the Watcher which may or may not have a maxOperationTime (but if it does, it takes precedent).